releaser/internal/versions/versions.go

```diff
@@ -0,0 +1,432 @@
+	vCfg, err := readVersioningFile(versioningFilename)
+	if err != nil {
+		return ModuleVersioning{}, fmt.Errorf("Error reading versioning file %v: %v", versioningFilename, err)
+	}
+
+	modSetMap, err := vCfg.buildModuleSetsMap()
+	if err != nil {
+		return ModuleVersioning{}, fmt.Errorf("Error building module set map for NewModuleVersioning: %v", err)
+	}
+
+	modInfoMap, err := vCfg.buildModuleMap()
+	if err != nil {
+		return ModuleVersioning{}, fmt.Errorf("Error building module info map for NewModuleVersioning: %v", err)
+	}
+
+	modPathMap, err := vCfg.BuildModulePathMap(repoRoot)
+	if err != nil {
+		return ModuleVersioning{}, fmt.Errorf("Error building module path map for NewModuleVersioning: %v", err)
+	}
+
+	return ModuleVersioning{
+		ModSetMap:  modSetMap,
+		ModPathMap: modPathMap,
+		ModInfoMap: modInfoMap,
+	}, nil
+}
+
+// MockModuleVersioning creates a ModuleVersioning struct for testing purposes.
+func MockModuleVersioning(modSetMap ModuleSetMap, modPathMap ModulePathMap) (ModuleVersioning, error) {
+	vCfg := versionConfig{
+		ModuleSets:      modSetMap,
+		ExcludedModules: []ModulePath{},
+	}
+
+	modInfoMap, err := vCfg.buildModuleMap()
+	if err != nil {
+		return ModuleVersioning{}, fmt.Errorf("error building module map: %v", err)
+	}
+
+	return ModuleVersioning{
+		ModSetMap:  modSetMap,
+		ModPathMap: modPathMap,
+		ModInfoMap: modInfoMap,
+	}, nil
+}
+
+// ModuleSetRelease contains info about a specific set of modules in the versioning file to be updated.
+type ModuleSetRelease struct {
+	ModuleVersioning
+	ModSetName string
+	ModSet     ModuleSet
+	TagNames   []ModuleTagName
+}
+
+// NewModuleSetRelease returns a ModuleSetRelease struct by specifying a specific set of modules to update.
+func NewModuleSetRelease(versioningFilename, modSetToUpdate, repoRoot string) (ModuleSetRelease, error) {
+	modVersioning, err := NewModuleVersioning(versioningFilename, repoRoot)
+	if err != nil {
+		return ModuleSetRelease{}, fmt.Errorf("unable to load baseVersionStruct: %v", err)
+	}
+
+	// get new version and mod tags to update
+	modSet, exists := modVersioning.ModSetMap[modSetToUpdate]
+	if !exists {
+		return ModuleSetRelease{}, fmt.Errorf("could not find module set %v in versioning file", modSetToUpdate)
+	}
+
+	// get tag names of mods to update
+	tagNames, err := modulePathsToTagNames(
+		modSet.Modules,
+		modVersioning.ModPathMap,
+		repoRoot,
+	)
+	if err != nil {
+		return ModuleSetRelease{}, fmt.Errorf("could not retrieve tag names from module paths: %v", err)
+	}
+
+	return ModuleSetRelease{
+		ModuleVersioning: modVersioning,
+		ModSetName:       modSetToUpdate,
+		ModSet:           modSet,
+		TagNames:         tagNames,
+	}, nil
+
+}
+
+// MockModuleSetRelease creates a ModuleSetRelease struct for testing purposes.
+func MockModuleSetRelease(modSetMap ModuleSetMap, modPathMap ModulePathMap, modSetToUpdate string, repoRoot string) (ModuleSetRelease, error) {
```

Comment:
This should move to a `versions_test.go` file so that it isn't exported for non-test use.

Reply:
I had several helper functions needed for testing that are likely to be shared between different subcommand packages. For example, I have the MockModuleVersioning() and WriteGoModFiles() funcs that will be used by the testing suites for `common`, `verify`, `prerelease`, and `tag`.

**I'm not sure which is the best approach for placing the files:**

1. If the funcs are moved to a `_test.go` file in common, they won’t be exported to the verify, prerelease, and tag subcommands.
2. If the funcs are moved into a separate package (such as `commontest`) that only provides exported test funcs, they are not usable by the `common` package, since `commontest` imports `common`, and then the tests for `common` would import `commontest`, so there would be a cycle.
3. If the funcs stay in the pkg `common`, they will be exported alongside all the “regular” funcs in the `common` package. (current implementation)